[feat] Control Flow Integrity (CFI) Features

mhatrevi opened the pull request:
This PR contains the following changes:

 1. CFI library with following features: 
        a. CFI counter to detect out-of-flow code execution. 
        b. XOR based Random number generator. 
        c. Procedural macro to add random delays before and after decorated functions and to detect out-of-flow code execution. 
        d. Variable launder function for assisting with duplicate checks.

 3. Enablement of ROM code with CFI features.

mhatrevi committed:
[feat] Control Flow Integrity (CFI) Features

 This PR contains the following changes:

 1. CFI library with following features:
	a. CFI counter to detect out-of-flow code execution.
	b. XOR based Random number generator.
	c. Procedural macro to add random delays before and after decorated functions
	   and to detect out-of-flow code execution.
	d. Variable launder function for assisting with duplicate checks.

 2. Enablement of ROM code with CFI features.

fdamato reviewed (commented):
This patch seems to have a very big impact on the debuggability of the code, making it in my opinion much more complex. From the meeting, it wasn`t really clear to me how these changes truly help. Would it possible to include some documentation which would clarify that ?

JohnTraverAmd commented on cfi/lib/src/cfi_ctr.rs:
```diff
@@ -0,0 +1,204 @@
+
+            let (new, overflow) = int.val.overflowing_add(1);
+            if overflow {
+                cfi_panic(CfiPanicInfo::CounterOverflow);
```
Where do we ensure that we never init the counter to a value where the overflow will happen naturally?

mhatrevi replied:
It's done at the start of rom_entry function in rom/dev/src/main.rs. The counter value that we start with is 0xA5A5A5A5 (0 XORed 0xA5A5A5A5).

JohnTraverAmd commented on cfi/derive/src/lib.rs:
```diff
@@ -0,0 +1,79 @@
+
+    wrapper_fn.block.stmts.clear();
+    wrapper_fn.block.stmts = parse_quote!(
+        let saved_ctr = caliptra_cfi_lib::CfiCounter::increment();
```
Can you describe the attack that you are defending with the counter increment and decrement?

mhatrevi replied:
The attack we are trying to prevent is of a malicious actor causing a glitch and skipping a CFI decorated function.
I discussed this with Marius today and we agreed that the way this is being done is not helpful from a glitch protection perspective; a more useful approach would be to embed either the inc/dec operation in the CFI decorated function and the corresponding opposite operation in the wrapper function. This will result in a counter mismatch should a CFI decorated function be skipped. This change will come as a separate PR.

mschilder123 commented on cfi/lib/src/cfi_ctr.rs:
```diff
@@ -0,0 +1,204 @@
+            if !val1.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+            if !val1.is_valid() {
```
val2, I presume?

mhatrevi replied:
Fixed.

mschilder123 commented on cfi/lib/tests/test_derive.rs:
```diff
@@ -0,0 +1,114 @@
+fn test_rand() {
+    let reg = Xoshiro128Reg;
+    let expected_rand_num: [u32; 10] = [
+        11520, 0, 5927040, 70819200, 2031721883, 1637235492, 1287239034, 3734860849, 3729100597,
```
Where do these KAT values come from?

mhatrevi replied:
I used the same reference implementation used in the code: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs

mschilder123 commented on cfi/derive/src/lib.rs:
```diff
@@ -0,0 +1,79 @@
+    File contains CFI procedural macro.
+
+References:
+    https://tf-m-user-guide.trustedfirmware.org/design_docs/tfm_physical_attack_mitigation.html
```
I do not see counter decrements in their suggestions. Not sure a counter increment/decrement wrapper achieves actual CFI monitoring? Entire calls can still be skipped afaict.

mhatrevi replied:
Based on the discussion we had today, we'll embed either the inc/dec operation in the CFI decorated function and corresponding opposite operation in the wrapper function. This will come as a separate PR.

RonaldAi commented on rom/dev/src/main.rs:
```diff
@@ -53,6 +54,13 @@ Running Caliptra ROM ...
 pub extern "C" fn rom_entry() -> ! {
     cprintln!("{}", BANNER);
 
+    if !cfg!(feature = "no-cfi") {
```
in other places this feature is called "no_cfi", here it is "no-cfi"

mhatrevi replied:
Typo; fixed.

korran commented on cfi/lib/src/xoshiro.rs:
```diff
@@ -0,0 +1,235 @@
+        unsafe {
+            #[cfg(feature = "cfi-test")]
+            {
+                core::ptr::read_volatile(&S0 as *const u32)
```
While volatile? These values aren't going to change behind the back of the compiler, are they?

korran replied:
Fixed.

korran commented on rom/dev/src/main.rs:
```diff
@@ -54,6 +55,13 @@ Running Caliptra ROM ...
 
+    if !cfg!(feature = "no-cfi") {
+        cprintln!("[state] CFI Enabled");
+        CfiCounter::reset();
```
CfiCounter::reset() inidrectly calls Trng::assume_init(), which is a bug (you can't seed from a trng that hasn't been in initialized yet).

korran replied:
Fixed.

korran commented on cfi/lib/src/cfi_ctr.rs:
```diff
@@ -0,0 +1,204 @@
+}
+
+/// CFI counter
+pub enum CfiCounter {}
```
nitpick: What's the point of making this a type? Shouldn't these just be top level functions in the module?

korran replied:
Example use:

```
// nitpick: cfi_counter would be clearer than cfi_ctr.
use cfi::cfi_ctr;

cfi_ctr::reset();
cfi_ctr::corrupt();
```

mhatrevi replied:
We'll make this change as part of the asym incr/decr proc. macro changes.

korran committed:
Make Xoshiro128 a normal rust type, and remove call to Trng::assume_init()

The latter fixes a serious bug with the csrng not being initialized
before we seed the xorshiro prng.

korran commented on cfi/lib/src/cfi_ctr.rs:
```diff
@@ -0,0 +1,223 @@
+    pub fn delay() {
+        #[cfg(all(target_arch = "riscv32", feature = "cfi", feature = "cfi-counter"))]
+        unsafe {
+            let cycles = prng().next() % 256;
```
Nitpick: Can you move these first two lines outside the unsafe block? It makes it easier for reviewers to be confident that prng().next() isn't an unsafe method with additional safety requirements.

mhatrevi replied:
Will do it as part of the asymmetrical increment and decrement calls in the proc macro.

korran commented on rom/dev/src/flow/cold_reset/fw_processor.rs:
```diff
@@ -170,10 +173,14 @@ impl FirmwareProcessor {
-        if let Some(result) = ImageManifest::read_from(slice.as_bytes()) {
+        let opt = ImageManifest::read_from(slice.as_bytes());
+        let result = opt.is_some();
+        if cfi_launder(result) {
```
I'm skeptical this is accomplishing much; we really want the code-gen to be executing the opt.is_some() line multiple times.

mhatrevi replied:
I'll update this as part of the asymm. incr/decr proc. macro changes.

korran commented on rom/dev/src/flow/update_reset.rs:
```diff
@@ -80,7 +82,11 @@ impl UpdateResetFlow {
         Self::populate_data_vault(venv.data_vault, info);
 
-        Self::load_image(&manifest, recv_txn)?;
+        Self::load_image(&manifest, &mut recv_txn)?;
```
What's the purpose of this change?

korran replied:
(cfi_impl_fn macro doesn't support mut args).

korran commented on rom/dev/src/pcr.rs:
```diff
@@ -52,111 +56,113 @@ impl PcrExtender<'_> {
+    ///
+    /// * `env` - ROM Environment
+    #[cfg_attr(not(feature = "no-cfi"), cfi_impl_fn)]
+    pub(crate) fn extend_pcrs(
```
(reverted)

korran committed:
Merge branch 'main' into vmhatre/cfi
